exp/internal/controllers/machinepool_controller.go

```diff
@@ -253,41 +253,64 @@ func (r *MachinePoolReconciler) reconcile(ctx context.Context, cluster *clusterv
 	return res, kerrors.NewAggregate(errs)
 }
 
-func (r *MachinePoolReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster, mp *expv1.MachinePool) error {
-	if ok, err := r.reconcileDeleteExternal(ctx, mp); !ok || err != nil {
+// reconcileDelete delete machinePool related resources.
+func (r *MachinePoolReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster, machinePool *expv1.MachinePool) error {
+	if ok, err := r.reconcileDeleteExternal(ctx, machinePool); !ok || err != nil {
 		// Return early and don't remove the finalizer if we got an error or
 		// the external reconciliation deletion isn't ready.
 		return err
 	}
 
-	if err := r.reconcileDeleteNodes(ctx, cluster, mp); err != nil {
-		// Return early and don't remove the finalizer if we got an error.
-		return err
+	// check nodes delete timeout passed.
+	if !r.isMachinePoolNodeDeleteTimeoutPassed(machinePool) {
+		if err := r.reconcileDeleteNodes(ctx, cluster, machinePool); err != nil {
+			// Return early and don't remove the finalizer if we got an error.
+			return err
+		}
 	}
 
-	controllerutil.RemoveFinalizer(mp, expv1.MachinePoolFinalizer)
+	controllerutil.RemoveFinalizer(machinePool, expv1.MachinePoolFinalizer)
 	return nil
 }
 
-func (r *MachinePoolReconciler) reconcileDeleteNodes(ctx context.Context, cluster *clusterv1.Cluster, machinepool *expv1.MachinePool) error {
-	if len(machinepool.Status.NodeRefs) == 0 {
+// reconcileDeleteNodes delete the cluster nodes.
+func (r *MachinePoolReconciler) reconcileDeleteNodes(ctx context.Context, cluster *clusterv1.Cluster, machinePool *expv1.MachinePool) error {
+	if len(machinePool.Status.NodeRefs) == 0 {
 		return nil
 	}
 
+	if r.Tracker == nil {
+		return errors.New("Cannot establish cluster client to delete nodes")
+	}
```

Comment:
I realize this is not the only place in the MachinePool controller where we check if the tracker is nil, but I'm a bit confused why we do it. The other place is in watchClusterNodes, where we just return nil if the Tracker is not set.

I assume it is simply not a valid case that the tracker is not set. Can we get rid of both checks that we have today and instead just have one in SetupWithManager (and clearly state there that this is a setup error)

Reply:
That seems pretty strange. I think we should just setup a ClusterCacheTracker in our unit tests instead of having handling like this in our production code. But okay let's keep it as is, it won't actually happen at regula runtime, so it's not really a problem to check